exercises/bowling/example.exs

```diff
@@ -0,0 +1,106 @@
+defmodule BowlingGame do
+  defstruct scores: [[0,0], [0,0], [0,0], [0,0], [0,0], [0,0], [0,0], [0,0], [0,0], [0,0], [0,0], [0,0]], current_frame: 1, roll_in_frame: 1
+end
+
+defmodule Bowling do
+  def new do
+    {:ok, game} = Agent.start_link(fn -> %BowlingGame{} end)
+    game
+  end
+
+  def roll(_, roll) when roll < 0 do
+    "Pins must have a value from 0 to 10"
+  end
+
+  def roll(_, roll) when roll > 10 do
+    "Pins must have a value from 0 to 10"
+  end
+
+  def roll(game, roll) do
+    game_info = get_game_info(game)
+    updates = update_score(game_info.roll_in_frame, game_info, roll)
+    cond do
+      valid_updates(updates) ->
+        Agent.update(game, fn(game_struct) -> updates end)
+      true ->
+        "Pin count exceeds pins on the lane" 
+    end
+  end
+
+  defp update_score(1, game_info, roll) do
+    current_frame = game_info.current_frame
+    scores = List.replace_at(game_info.scores, current_frame - 1, [roll, 0])
+    cond do
+      roll == 10 ->
+        %{ game_info | current_frame: current_frame + 1, scores: scores }
+      true ->
+        %{ game_info | roll_in_frame: 2, scores: scores }
+    end
+  end
+
+  defp update_score(2, game_info, roll) do
+    current_frame = game_info.current_frame
+    old_scores = game_info.scores
+    old_sum_lists = Enum.at(old_scores, current_frame - 1)
+    new_sum_lists = List.replace_at(old_sum_lists, 1, roll)
+    new_scores = List.replace_at(old_scores, current_frame - 1, new_sum_lists)
+    %{ game_info | scores: new_scores, roll_in_frame: 1,
+                   current_frame: current_frame + 1 }
+  end
+
+  defp valid_updates(updates) do
+    Enum.all?(updates.scores, fn(frame) ->
+      Enum.reduce(frame, 0, fn(score, acc) -> score + acc end) <= 10
+    end)
+  end
+
+  def score(game) do
+    game_struct = Agent.get(game, fn(game_struct) -> game_struct end)
+    if game_struct.current_frame < 11 do
+      "Score cannot be taken until the end of the game"
+    else
+      parse_scores(game_struct.scores)
+    end
+  end
+
+  defp get_game_info(game) do
+    Agent.get(game, fn(game_struct) -> game_struct end)
+  end
+
+  defp parse_scores(scores) do
+    scores
+      |> score_frames
+      |> sum_list
+  end
+
+  defp score_frames(scores) do
+    Enum.map((0..9), fn(idx) ->
+      current_frame = Enum.at(scores, idx)
+      next_frame = Enum.at(scores, idx + 1, [0,0])
+      cond do
+        strike(current_frame) ->
+          if strike(next_frame) do
+            sum_list(current_frame) + sum_list(next_frame) + (Enum.at(scores, idx + 2, [0, 0]) |> Enum.at(0))
+          else
+            sum_list(Enum.at(scores, idx)) + sum_list(Enum.at(scores, idx + 1, [0,0]))
+          end
+        spare(current_frame) ->
+          10 + Enum.at(next_frame, 0)
+        true ->
+          sum_list(current_frame)
+      end
+    end)
+  end
```

Comment:
I'd definitely split this out into several functions, probably one to calculate strike/spare/other